File: Stack/Haskell/Stack.hs
{- |
Module      :   Stack.hs
Description :   Implementation of a Stack
Copyright   :   (c) Thomas Lang, 2014
License     :   BSD3

Stability   :   stable
Portability :   portable

This module implements a classical Stack with an abstract
data type. It supports all typical functions, like "push",
"pop" and "isEmpty".
Furthermore, it gives the opportunity to create a stack from
a list.
 -}

module Stack (  emptyStack,     -- ^ constant function for Empty Stack
                isEmpty,        -- ^ checks if stack is empty or not
                fromList,       -- ^ creates a Stack from list
                push,           -- ^ Pushes an element to a Stack
                pop,            -- ^ Pops the Stack
                top             -- ^ Returns the top element of a Stack
             ) where



-- --------------------------------------------------------
-- Abstract data type
-- --------------------------------------------------------
data STACK a = EmptyStack
             | S a (STACK a)
             deriving (Ord)


-- --------------------------------------------------------
-- instance for checking the equality of two Stacks
-- --------------------------------------------------------
instance (Eq a) => Eq (STACK a) where
    x == y  =  x `equalsTree` y


-- -------------------------------------------------------
-- instance for pretty-printing a Stack
-- -------------------------------------------------------
instance (Show a) => Show (STACK a) where
    show EmptyStack = "EmptyStack"
    show (S x y)    = "Stack " ++ show x ++ " ( " ++ show y ++ ")"


-- --------------------------------------------------------
-- Constant function, creates empty Stack
-- --------------------------------------------------------
emptyStack :: STACK a
emptyStack = EmptyStack
 

-- --------------------------------------------------------
-- Checks whether a Stack is empty or not
-- --------------------------------------------------------
isEmpty :: STACK a -> Bool
isEmpty EmptyStack = True
isEmpty _          = False


-- --------------------------------------------------------
-- private functions that checks the equality
-- --------------------------------------------------------
equalsTree :: (Eq a) => STACK a -> STACK a -> Bool
equalsTree EmptyStack EmptyStack       = True
equalsTree (S x y)    (S u v) | x == u = equalsTree y v
equalsTree _           _               = False


-- --------------------------------------------------------
-- Creates a Stack from a list
-- --------------------------------------------------------
fromList :: [a] -> STACK a
fromList []     = EmptyStack
fromList (x:xs) = S x (fromList xs )


-- --------------------------------------------------------
-- Pushes a new element to a stack
-- --------------------------------------------------------
push :: a -> STACK a -> STACK a
push x EmptyStack = S x EmptyStack
push x s          = S x s


-- --------------------------------------------------------
-- Pops the stack, so it removes the top element
-- --------------------------------------------------------
pop :: STACK a -> Maybe (STACK a)
pop EmptyStack = Nothing
pop (S x y)    = Just y


-- -------------------------------------------------------
-- Returns the top element of the stack
-- -------------------------------------------------------
top :: STACK a -> Maybe a
top EmptyStack = Nothing
top (S x y)    = Just x

File: Stack/C/stack.c
/* Stack
 * (C) Thomas Lang, 2014
 *
 * This module implements functions and methods
 * used for working with a typical Stack data
 * structure, explanation can be found in the header.
 *
 * This code is licensed under the BSD3 license.
 */

#include <stdio.h>
#include <stdlib.h>

#include "stack.h"

//----------------------------------------
Node *newNode( int value )
{
    Node *node  = malloc( sizeof( Node ));
    node->value = value;
    node->next  = NULL;
    return node;
}

//---------------------------------------
int getSize( Stack *stack )
{
    return stack->size;
}


//---------------------------------------
Stack *newStack()
{
    Stack *stack = malloc( sizeof( Stack ));
    stack->sp    = NULL;
    stack->size  = 0;
    return stack;
}


//--------------------------------------
int isEmpty( Stack *stack )
{
    if( stack->sp == NULL )
        return 0;
    return (-1);
}

//-------------------------------------
void push( Stack *stack, int value )
{
    Node *node = newNode( value );
    if( isEmpty( stack ) == 0 )
    {
        stack->sp  = node;
    }
    else
    {
        node->next = stack->sp;
        stack->sp  = node;
    }
    stack->size++;
}

//------------------------------------
Node *pop( Stack *stack )
{
    if( isEmpty( stack ) == 0 )
    {
        printf( "Cannot pop from empty Stack!\n" );
        return NULL;
    }
    Node *tmp = stack->sp;
    stack->sp = stack->sp->next;
    stack->size--;
    return tmp;
}

//------------------------------------
Node *top( Stack *stack )
{
    return stack->sp;
}


//-------------------------------------------
void printStack( Stack *stack )
{
    if( isEmpty( stack ) == 0)
        printf( "(empty stack)\n" );
    else
    {
        printf( "---> %d\n", stack->sp->value );
        Node *top = stack->sp->next;
        while( top != NULL )
        {
            printf( "     %d\n", top->value );
            top = top->next;
        }
        printf( "\n" );
    }
}
